sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -884,28 +885,81 @@ public PDone expand(PCollection<byte[]> input) {
           getPartitionKey() == null || (getPartitioner() == null),
           "only one of either withPartitionKey() or withPartitioner() is possible");
       checkArgument(getAWSClientsProvider() != null, "withAWSClientsProvider() is required");
+      producerConfiguration(); // verify Kinesis producer configuration can be build
 
       input.apply(ParDo.of(new KinesisWriterFn(this)));
       return PDone.in(input.getPipeline());
     }
 
-    private static class KinesisWriterFn extends DoFn<byte[], Void> {
+    private KinesisProducerConfiguration producerConfiguration() {
+      Properties props = getProducerProperties();
+      if (props == null) {
+        props = new Properties();
+      }
+      return KinesisProducerConfiguration.fromProperties(props);
+    }
 
+    private static class KinesisWriterFn extends DoFn<byte[], Void> {
       private static final int MAX_NUM_FAILURES = 10;
+      /** Usage count of static, shared Kinesis producer. */
+      private static final AtomicInteger producerRefCount = new AtomicInteger();
+
+      /** Static, shared Kinesis producer. */
+      private static IKinesisProducer producer;
 
       private final KinesisIO.Write spec;
-      private static transient IKinesisProducer producer;
+
       private transient KinesisPartitioner partitioner;
       private transient LinkedBlockingDeque<KinesisWriteException> failures;
       private transient List<Future<UserRecordResult>> putFutures;
 
       KinesisWriterFn(KinesisIO.Write spec) {
         this.spec = spec;
-        initKinesisProducer();
+      }
+
+      /**
+       * Initialize statically shared Kinesis producer if required and count usage.
+       *
+       * <p>NOTE: If there is, for whatever reasons, another instance of a {@link KinesisWriterFn}
+       * with different producer properties or even a different implementation of {@link
+       * AWSClientsProvider}, these changes will be silently discarded in favor of an existing
+       * producer instance.
+       */
+      @SuppressFBWarnings("JLM_JSR166_UTILCONCURRENT_MONITORENTER")
+      private void setupSharedProducer() {
+        synchronized (producerRefCount) {
+          if (producer == null) {
+            producer =
+                spec.getAWSClientsProvider().createKinesisProducer(spec.producerConfiguration());
+            producerRefCount.set(0);
+          }
+        }
+        producerRefCount.incrementAndGet();
```

Comment:
Wouldn't we have 2 memory fences here? Although I can not come up with a code path here, I could imagine someone adding some code in setup before calling into setupSharedProducer. If that throws, we could end up with setting refCount to 0 in the synchronised block here, get suspended before line 937, teardown on that other instance will called, resulting in refCount to be -1. From there on we will be off by one. Does this make any sense?

Reply:
@adude3141 I suppose there is a risk in general that we're getting out of sync if the code is changed in "careless" ways, agreed. As we're stuck with the pair of `setup` and `teardown` I don't see a good way to prevent this, but I'm open for suggestions!

Reply:
Sorry for being unclear. I am more concerned about the 'publishing' of inconsistent state after releasing the lock/synchronized block in line 936. producerRecCount is 0, where it should be 1.

I d say an easy fix could be to simply pull that producerRefCount.incrementAndGet() (line 937) into the synchronised block. As we aquire the lock/synchronize anyway, this will not hurt. Of course, we could consider switching to plain int then as a counter, as it is accessed in teardown also behind the lock.

Does this make more sense?

Reply:
Thanks @adude3141 , in deed an oversight 👍  Fixed!

Reply:
So, thanks a lot for your contribution.

Reply:
@mosche, what about switching from `AtomicInteger` to `Integer` as @adude3141 suggested ? Indeed, this counter is always accessed in a synchronised block, there is no need to add extract sync mechanism. And by the way it will also allow to remove the `SuppressWarning`

Reply:
That would require to introduce a new static object that can be used as a lock for synchronization as this can't be done on the immutable integer. Also, that integer needs to be volatile as accessed from multiple threads. I'd honestly prefer keeping the atomic int for simplicity, but I absolutely don't have a strong opinion here. 
Alternatively something like `private static final int[] refCount = {0}` could be used. 
Let me know what your preference is ...

Reply:
True, incrementing the Integer means creating another instance of it, so yes, to maintain the lock we need to create an `Object` for the sole lock purpose. Which does not seem strange to me. So we could use a simple int for the counter. I don't think it needs to be volatile because it is always accessed from a synchronized block and these block invocations will create a happens-before relationship at the lock level (the Object lock) that makes the changes visible to other threads that lock on the same Object. I think this solution is clearer for maintenance, because it avoids people wondering why we synchronise and use the `AtomicInteger`.

Reply:
i don't think maintenance was ever a consideration in this IO before 🤯  but you're certainly right @echauchot and i'm falling for lazy decisions .. 😳 
anyways, having another look at this, i also pulled `flushSync` and `destroy` out of the synchronized block to prevent new writer instances from waiting much longer than necessary, though unlikely. 
also note, the if clause around  `flushSync` is obsolete and was removed.

Reply:
@adude3141 @echauchot  Would you have time to review https://github.com/apache/beam/pull/16077 as well?
At least parts is related to this as well, see here https://github.com/apache/beam/pull/16077/files#diff-affa5b955197f29e96d8657324205e5105359dc6672cfffc16d7e9ddfd55047a

Reply:
Your right of course, thanks!